File: leetcode/剑指 Offer 系列/面试题 35. 复杂链表的复制.js
/**
 * Difficulty:
 * Medium
 *
 * Desc:
 * 请实现 copyRandomList 函数，复制一个复杂链表。在复杂链表中，每个节点除了有一个 next 指针指向下一个节点，还有一个 random 指针指向链表中的任意节点或者 null
 *
 * 同：https://leetcode-cn.com/problems/copy-list-with-random-pointer/
 */

/**
 * // Definition for a Node.
 * function Node(val,next,random) {
 *    this.val = val;
 *    this.next = next;
 *    this.random = random;
 * };
 */
/**
 * @param {Node} head
 * @return {Node}
 */
var copyRandomList = function(head) {
  if (!head) return null

  let node = head
  while (node) {
    const newNode = new Node(node.val, node.next, null)
    node.next = newNode
    node = newNode.next
  }

  node = head
  while (node && node.next) {
    if (node.random) node.next.random = node.random.next
    node = node.next.next
  }

  node = head
  const result = node.next
  while (node && node.next) {
    const next = node.next
    node.next = next.next
    node = next
  }

  return result
}

File: leetcode/剑指 Offer 系列/面试题 35. 复杂链表的复制.cs
/**
 * Difficulty:
 * Medium
 *
 * Desc:
 * 请实现 copyRandomList 函数，复制一个复杂链表。
 * 在复杂链表中，每个节点除了有一个 next 指针指向下一个节点，还有一个 random 指针指向链表中的任意节点或者 null。
 *
 * 示例 1：
 * 输入：head = [[7,null],[13,0],[11,4],[10,2],[1,0]]
 * 输出：[[7,null],[13,0],[11,4],[10,2],[1,0]]
 *
 * 示例 2：
 * 输入：head = [[1,1],[2,1]]
 * 输出：[[1,1],[2,1]]
 *
 * 示例 3：
 * 输入：head = [[3,null],[3,0],[3,null]]
 * 输出：[[3,null],[3,0],[3,null]]
 *
 * 示例 4：
 * 输入：head = []
 * 输出：[]
 * 解释：给定的链表为空（空指针），因此返回 null。
 *
 * 提示：
 * -10000 <= Node.val <= 10000
 * Node.random 为空（null）或指向链表中的节点。
 * 节点数目不超过 1000 。
 * 注意：本题与主站 138 题相同：https://leetcode-cn.com/problems/copy-list-with-random-pointer/
 */

/*
// Definition for a Node.
public class Node {
    public int val;
    public Node next;
    public Node random;

    public Node(){}
    public Node(int _val,Node _next,Node _random) {
        val = _val;
        next = _next;
        random = _random;
    }
}
*/

public class Solution {
    public Node CopyRandomList(Node head) {
        Node cur = head;
        while (cur != null) {
            Node next = cur.next;
            cur.next = new Node(cur.val);
            cur.next.next = next;
            cur = next;
        }

        cur = head;

        while (cur != null) {
            Node copyNext = cur.next;

            Node random = cur.random;
            if (random != null) copyNext.random = random.next;

            cur = copyNext.next;
        }

        Node res = null;
        while (head != null) {
            Node copyNext = head.next;
            Node realNext = copyNext.next;

            if (res == null) res = copyNext;
            if (realNext != null) copyNext.next = realNext.next;

            head.next = realNext;
            head = head.next;
        }

        return res;
    }
}
